unittests/rpc/handlers/BookOffersTest.cpp

```diff
@@ -0,0 +1,1327 @@
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <rpc/RPCHelpers.h>
+#include <rpc/common/AnyHandler.h>
+#include <rpc/ngHandlers/BookOffers.h>
+#include <util/Fixtures.h>
+#include <util/TestObject.h>
+
+#include <fmt/core.h>
+
+constexpr static auto ACCOUNT = "rf1BiGeXwwQoi8Z2ueFYTEXSwuJYfV2Jpn";
+constexpr static auto ACCOUNT2 = "rLEsXccBGNR3UPuPu2hUXPjziKC3qKSBun";
+
+constexpr static auto LEDGERHASH =
+    "4BC50C9B0D8515D3EAAE1E74B29A95804346C491EE1A95BF25E4AAB854A6A652";
+constexpr static auto INDEX1 =
+    "1B8590C01B0006EDFA9ED60296DD052DC5E90F99659B25014D08E1BC983515BC";
+constexpr static auto INDEX2 =
+    "E6DBAFC99223B42257915A63DFC6B0C032D4070F9A574B255AD97466726FC321";
+// 20 USD : 10 XRP
+constexpr static auto PAYS20USDGETS10XRPBOOKDIR =
+    "43B83ADC452B85FCBADA6CAEAC5181C255A213630D58FFD455071AFD498D0000";
+// 20 XRP : 10 USD
+constexpr static auto PAYS20XRPGETS10USDBOOKDIR =
+    "7B1767D41DBCE79D9585CF9D0262A5FEC45E5206FF524F8B55071AFD498D0000";
+// transfer rate x2
+constexpr static auto TRANSFERRATEX2 = 2000000000;
+
+using namespace RPCng;
+namespace json = boost::json;
+using namespace testing;
+
+class RPCBookOffersHandlerTest : public HandlerBaseTest
+{
+};
+
+struct ParameterTestBundle
+{
+    std::string testName;
+    std::string testJson;
+    std::string expectedError;
+    std::string expectedErrorMessage;
+};
+
+struct RPCBookOffersParameterTest
+    : public RPCBookOffersHandlerTest,
+      public WithParamInterface<ParameterTestBundle>
+{
+    struct NameGenerator
+    {
+        template <class ParamType>
+        std::string
+        operator()(const testing::TestParamInfo<ParamType>& info) const
+        {
+            auto bundle = static_cast<ParameterTestBundle>(info.param);
+            return bundle.testName;
+        }
+    };
+};
+
+TEST_P(RPCBookOffersParameterTest, CheckError)
+{
+    auto bundle = GetParam();
+    auto const handler = AnyHandler{BookOffersHandler{mockBackendPtr}};
+    runSpawn([&](boost::asio::yield_context yield) {
+        auto const output =
+            handler.process(json::parse(bundle.testJson), yield);
+        ASSERT_FALSE(output);
+        auto const err = RPC::makeError(output.error());
+        EXPECT_EQ(err.at("error").as_string(), bundle.expectedError);
+        EXPECT_EQ(
+            err.at("error_message").as_string(), bundle.expectedErrorMessage);
+    });
+}
+
+auto
+generateParameterBookOffersTestBundles()
+{
+    return std::vector<ParameterTestBundle>{
+        ParameterTestBundle{
+            "MissingTakerGets",
+            R"({
+                "taker_pays" : {
+                    "currency" : "USD",
+                    "issuer" : "rvYAfWj5gh67oV6fW32ZzP3Aw4Eubs59B"
+                    }
```

Comment:
minor: Again some strange formatting here and below in other json strings. Consider fixing :)